pkg/collector/corechecks/windows_event_log/check.go

```diff
@@ -0,0 +1,294 @@
+
+// The lower cased version of the `API SOURCE ATTRIBUTE` column from the table located here:
+// https://docs.datadoghq.com/integrations/faq/list-of-api-source-attribute-value/
+const sourceTypeName = "event viewer"
+
+// Check defines a check that reads the Windows Event Log and submits Events
+type Check struct {
+	// check
+	core.CheckBase
+	config *Config
+
+	fetchEventsLoopWaiter sync.WaitGroup
+	fetchEventsLoopStop   chan struct{}
+
+	includedMessages []*regexp.Regexp
+	excludedMessages []*regexp.Regexp
+
+	// event metrics
+	eventPriority agentEvent.EventPriority
+
+	// event log
+	session             evtsession.Session
+	sub                 evtsubscribe.PullSubscription
+	evtapi              evtapi.API
+	systemRenderContext evtapi.EventRenderContextHandle
+	userRenderContext   evtapi.EventRenderContextHandle
+	bookmarkSaver       *bookmarkSaver
+}
+
+// Run updates sender stats, restarts the subscription if it failed, and saves the bookmark.
+// The main event collection logic runs continuously in the background, not during Run().
+func (c *Check) Run() error {
+	sender, err := c.GetSender()
+	if err != nil {
+		return err
+	}
+	// Necessary for check stats to be calculated (number of events collected, etc)
+	// Since events are collected in the background, this will update stats with the
+	// count of events collected since the last Run() call.
+	defer sender.Commit()
+
+	// Start/Restart the subscription if it is not running
+	if !c.sub.Running() {
+		// starts the event collection in the background.
+		err := c.startSubscription()
+		if err != nil {
+			return err
+		}
+	}
+
+	// save/persist bookmark on each check run
+	// check.Cancel() may not be called (like in agent check command), which means
+	// if the events read are less than bookmark_frequency then a bookmark won't
+	// be saved before the process exits. saving here, too, gives us a good time periodic
+	// save/persist in addition to the count periodic bookmark_frequency option.
+	err = c.bookmarkSaver.saveLastBookmark()
+	if err != nil {
+		c.Warnf(err.Error())
+	}
+
+	return nil
+}
+
+func (c *Check) fetchEventsLoop(sender sender.Sender) {
+	defer c.fetchEventsLoopWaiter.Done()
+
+	// Always stop the subscription when the loop ends.
+	// The check will start the subscription and this loop again next time it runs.
+	defer c.sub.Stop()
+
+	// Save the bookmark at the end of the loop, regardless of the bookmarkFrequency
+	defer func() {
+		err := c.bookmarkSaver.saveLastBookmark()
+		if err != nil {
+			c.Warnf(err.Error())
+		}
+	}()
+
+	// Fetch new events
+	for {
+		select {
+		case <-c.fetchEventsLoopStop:
+			log.Debug("stopping subscription")
+			return
+		case events, ok := <-c.sub.GetEvents():
+			if !ok {
+				// The channel is closed, this indicates an error or that sub.Stop() was called
+				// Use sub.Error() to get the error, if any.
+				err := c.sub.Error()
+				if err != nil {
+					log.Errorf("event subscription stopped with error: %v", err)
```

Comment:
Would it be possible to return errors from `Run` as well, so they show up in the status page and reflect in the check status in the app metadata?